File: ex10/ex10_02/src/polyline.cpp
#include "polyline.hpp"
using namespace std;

void Polyline::add_point(double x, double y) {
  points.emplace_back(x, y);
}
double Polyline::get_x(size_t index) {
  if (index < this->points.size()) {
      return this->points[index].x;
  }
  throw std::out_of_range("Index out of range");
}
double Polyline::get_y(size_t index) {
  if (index < this->points.size()) {
      return this->points[index].y;
  }
  throw std::out_of_range("Index out of range");
}
double Polyline::get_distance(size_t i, size_t j) {
  if (i < 0 || i >= Polyline::get_size() || j < 0 || j >= Polyline::get_size()) {
    throw out_of_range("Invalid index");
  }
  double dx = this->get_x(j) - this->get_x(i);
  double dy = this->get_y(j) - this->get_y(i);
  return sqrt(dx * dx + dy * dy);
}
double Polyline::get_length() {
  double total_length = 0.0;
  for (int i = 1; i < this->Polyline::get_size(); ++i) {
    total_length += this->get_distance(i - 1, i);
  }
  return total_length;
}
size_t Polyline::get_size() const {
  return this->points.size();
};

File: ex10/ex10_02/src/polyline.py
class Polyline:
  def __init__(self, n):
    self.points = []
    self.num_points = n
    for i in range(n):
      self.points.append({'x': 0.0, 'y': 0.0})
  
  def get_distance(self, i, j):
    try:
      if (i < 0 or i >= self.num_points or j < 0 or j >= self
          .num_points):
        raise ValueError("Invalid index")
      if (i == j):
        return 0.0  
      dx = self.points[j]['x'] - self.points[i]['x']
      dy = self.points[j]['y'] - self.points[i]['y']
      return (dx * dx + dy * dy) ** 0.5
    except ValueError as err:
      print(err)
      return 0.0

  def get_length(self):
    if (self.num_points < 2):
      return 0.0
    else:
      totalLength = 0.0
      for i in range(1, self.num_points):
        totalLength += self.get_distance(i - 1, i)
      return totalLength

